File: apps/trace-propagation/sqs/java/src/main/java/example/Consumer.java
package example;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.SQSEvent;
import com.timgroup.statsd.NonBlockingStatsDClientBuilder;
import com.timgroup.statsd.StatsDClient;
import org.json.JSONObject;

import static example.Context.currentTraceId;
import static example.Context.runtime;

public class Consumer implements RequestHandler<SQSEvent, Void>{

  private static final StatsDClient statsd = new NonBlockingStatsDClientBuilder().hostname("localhost").build();

  @Override
  public Void handleRequest(SQSEvent event, Context context)
  {
    String traceId = currentTraceId();
    for (SQSEvent.SQSMessage message : event.getRecords()) {
      String payload = message.getBody();
      System.out.println("received sqs message " + payload);

      JSONObject parsed = new JSONObject(payload);
      statsd.recordDistributionValue("trace_context.propagated.sqs", 1, new String[]{
        "consumer_runtime:" + runtime,
        "producer_runtime:" + parsed.optString("runtime"),
        parsed.optString("trace_id").equals(traceId) ? "success:true" : "success:false",
        "transport:sqs",
      });
    }
    return null;
  }
}

File: go/sqs/consumer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"

	ddlambda "github.com/DataDog/datadog-lambda-go"
	"github.com/DataDog/sample-apps/go/sqs/internal"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
)

func consumer(ctx context.Context, event events.SQSEvent) (events.SQSEventResponse, error) {
	traceID := internal.TraceID(ctx)
	parent, _ := tracer.SpanFromContext(ctx)
	for _, record := range event.Records {
		span := tracer.StartSpan("sqs.record.consume", tracer.ChildOf(parent.Context()))
		{
			var msg internal.Message
			fmt.Printf("received sqs message %s\n", record.Body)
			json.Unmarshal([]byte(record.Body), &msg)
			ddlambda.Metric(
				"trace_context.propagated.sqs", 1,
				fmt.Sprintf("consumer_runtime:%s", internal.Runtime),
				fmt.Sprintf("producer_runtime:%s", msg.Runtime),
				fmt.Sprintf("success:%t", traceID == msg.TraceID),
				"transport:sqs",
			)
		}
		span.Finish()
	}
	return events.SQSEventResponse{}, nil
}

func main() {
	lambda.Start(ddlambda.WrapFunction(consumer, nil))
}
